tls/s2n_ktls.c

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#ifdef S2N_PLATFORM_SUPPORTS_KTLS
+    /*
+     * https://github.com/aws/s2n-tls/issues/3813
+     *
+     * _GNU_SOURCE is needed for resolving the constant SOL_TCP
+     * when building `tls/s2n_ktls.c`.
+     */
+    #ifndef _GNU_SOURCE
+        #define _GNU_SOURCE
+        #include <netinet/tcp.h>
+        #undef _GNU_SOURCE
+    #else
+        #include <netinet/tcp.h>
+    #endif
+
+    #include <linux/tls.h>
+    #include <sys/socket.h>
+
+    #include "error/s2n_errno.h"
+    #include "tls/s2n_connection.h"
+    #include "utils/s2n_result.h"
+
+    #define S2N_TLS_ULP_NAME      "tls"
+    #define S2N_TLS_ULP_NAME_SIZE sizeof(S2N_TLS_ULP_NAME)
```

Comment:
Putting these in the if/def complicates things unnecessarily. These headers will always exist, and if the defines exisitng doesn't hurt anything they should always exist. Only conditionally compile the parts that might not be there.